[ao] Adding model report detector base class and implemented detectors

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79792
* #79752
* #79595
* __->__ #79671

Summary: the goal is to add a base class to the model report detectors
so that they can contain a lot more specific information compared to the
primary model report class related to the observers and where they are
inserted etc.

Since this is just a base class, the testing will be with the
implemenations of the classes that derive from the base class

The two current detector methods were turned into Detector classes and
the tests were modified to reflect this, but the same functionality was
tested.

As a result, _detector.py was changed to detector.py

Test Plan: python test/test_quantization.py TestFxModelReportDetector
python test/test_quantization.py TestFxModelReportDetectDynamicStatic

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao] Adding model report detector base class and implemented detectors

Summary: the goal is to add a base class to the model report detectors
so that they can contain a lot more specific information compared to the
primary model report class related to the observers and where they are
inserted etc.

Since this is just a base class, the testing will be with the
implemenations of the classes that derive from the base class

The two current detector methods were turned into Detector classes and
the tests were modified to reflect this, but the same functionality was
tested.

As a result, _detector.py was changed to detector.py

Test Plan: python test/test_quantization.py TestFxModelReportDetector
python test/test_quantization.py TestFxModelReportDetectDynamicStatic

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Adding model report detector base class and implemented detectors"


Summary: the goal is to add a base class to the model report detectors
so that they can contain a lot more specific information compared to the
primary model report class related to the observers and where they are
inserted etc.

Since this is just a base class, the testing will be with the
implemenations of the classes that derive from the base class

The two current detector methods were turned into Detector classes and
the tests were modified to reflect this, but the same functionality was
tested.

As a result, _detector.py was changed to detector.py

Test Plan: python test/test_quantization.py TestFxModelReportDetector
python test/test_quantization.py TestFxModelReportDetectDynamicStatic

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -0,0 +1,406 @@
+from torch.ao.quantization.fx.graph_module import GraphModule
+from torch.ao.quantization.observer import ObserverBase
+from torch.ao.quantization.qconfig import QConfig
+from torch.nn.qat.modules.conv import _ConvNd as QatConvNd
```
maybe write down concrete modules instead of _ConvNd, e.g. nn.qat.Conv1d, nn.qat.Conv2d, nn.qat.Conv3d

jerryzh168 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -0,0 +1,406 @@
+from torch.ao.quantization.observer import ObserverBase
+from torch.ao.quantization.qconfig import QConfig
+from torch.nn.qat.modules.conv import _ConvNd as QatConvNd
+from torch.nn.qat.modules.linear import Linear as QatLinear
```
the exposed api is torch.nn.qat.Linear I think

typically, what we do is:
```
# import
import torch.nn.qat as nnqat

# use
nnqat.Linear, nnqat.Conv1d, nnqat.Conv2d etc..
```

jerryzh168 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -0,0 +1,406 @@
+    }
+
+    def __init__(self):
+        super(PerChannelDetector, self).__init__()
```
nit: I think we are mostly using `super().__init__()` now

jerryzh168 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -0,0 +1,406 @@
+        super(PerChannelDetector, self).__init__()
+
+        # store the backend information
+        self.backend_chosen = torch.backends.quantized.engine
```
I'm wondering if this should be an argument, since the backend that user wants to use to in the end (target backend to use in production) may not be the same as `torch.backends.quantized.engine` (current quantized backend selected when user is preparing the model)

vspenubarthi committed:
Update on "[ao] Adding model report detector base class and implemented detectors"


Summary: the goal is to add a base class to the model report detectors
so that they can contain a lot more specific information compared to the
primary model report class related to the observers and where they are
inserted etc.

Since this is just a base class, the testing will be with the
implemenations of the classes that derive from the base class

The two current detector methods were turned into Detector classes and
the tests were modified to reflect this, but the same functionality was
tested.

As a result, _detector.py was changed to detector.py

Test Plan: python test/test_quantization.py TestFxModelReportDetector
python test/test_quantization.py TestFxModelReportDetectDynamicStatic

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -0,0 +1,408 @@
+from torch.ao.quantization.observer import ObserverBase
+from torch.ao.quantization.qconfig import QConfig
+
+ABC: Any = ABCMeta(str("ABC"), (object,), {})
```
This seems unnecessarily complex. Why not just extend `abc.ABC`?

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -0,0 +1,408 @@
+        raise NotImplementedError("No observers are inserted in the PerChannelDetector.")
+
+
+    def __detect_per_channel_helper(self, model: nn.Module, per_channel_info: Dict):
```
nit: I think we usually just use 1 underscore for private methods

vspenubarthi replied:
Just checked locally and it has to be two. when i started the methods with only 1 underscore, they were still accessible publically when I tried to call them after creating an instance.

vspenubarthi replied:
looked up docs and I stand corrected, guess it should be one, i will fix

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -0,0 +1,408 @@
+            # however, we want to also recommend they add a dynamic quantize per tensor right if this change is made
+            if (
+                module_info["pre_observer_data_dist"] == self.DEFAULT_NON_STATIONARY
+                and module_info["post_observer_data_dist"] == self.DEFAULT_STATIONARY
```
I wonder if we should make these dict keys constant as well (like `DEFAULT_STATIONARY`) so we're not duplicating them everywhere? If there's a typo somewhere the result could be hard to debug

andrewor14 reviewed (approved):
Looks great!

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -0,0 +1,408 @@
+        pass
+
+    @abstractmethod
+    def generate_detector_report(self, model):
```
Can you add the return types here as well (and all other abstract methods)?

HDCharles commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -0,0 +1,408 @@
+        "onednn": set([nn.Linear, nn.Conv1d, nn.Conv2d, nn.Conv3d, nnqat.Linear, nnqat.Conv1d, nnqat.Conv2d, nnqat.Conv3d]),
+    }
+
+    def __init__(self, backend):
```
nit: as Jerry mentioned, I think its a good idea to take this as an argument, but it should probably be optional, we can fall back to the current set backend if this isn't set.

HDCharles commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -0,0 +1,408 @@
+        r""" returns the string name of this detector"""
+        return "dynamic_vs_static_detector"
+
+    def __generate_dict_info(self, model: GraphModule) -> Dict[str, Any]:
```
should this be one _ or two __ ?

vspenubarthi replied:
I think it has to be 2 for it to be private, just tested it and if it's only 1, it still pops up as a callable method when we create an instance.

vspenubarthi replied:
looked up docs and I stand corrected, guess it should be one, i will fix

vspenubarthi committed:
Update on "[ao] Adding model report detector base class and implemented detectors"


Summary: the goal is to add a base class to the model report detectors
so that they can contain a lot more specific information compared to the
primary model report class related to the observers and where they are
inserted etc.

Since this is just a base class, the testing will be with the
implemenations of the classes that derive from the base class

The two current detector methods were turned into Detector classes and
the tests were modified to reflect this, but the same functionality was
tested.

As a result, _detector.py was changed to detector.py

Test Plan: python test/test_quantization.py TestFxModelReportDetector
python test/test_quantization.py TestFxModelReportDetectDynamicStatic

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Adding model report detector base class and implemented detectors"


Summary: the goal is to add a base class to the model report detectors
so that they can contain a lot more specific information compared to the
primary model report class related to the observers and where they are
inserted etc.

Since this is just a base class, the testing will be with the
implemenations of the classes that derive from the base class

The two current detector methods were turned into Detector classes and
the tests were modified to reflect this, but the same functionality was
tested.

As a result, _detector.py was changed to detector.py

Test Plan: python test/test_quantization.py TestFxModelReportDetector
python test/test_quantization.py TestFxModelReportDetectDynamicStatic

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@pytorchbot merge -g
